File: 2.linked-list/single-linked-list/merge-sort/mergeSortGeneral.js
const mergeSort = function (arr) {
  if (arr.length <= 1) return arr;
  let mid = Math.floor(arr.length / 2);
  let right = [...arr];
  let left = right.splice(0, mid);

  return merge(mergeSort(left), mergeSort(right));
};

const merge = function (left, right) {
  const result = [];
  while (left.length && right.length) {
    if (left[0] < right[0]) {
      result.push(left.shift());
    } else {
      result.push(right.shift());
    }
  }
  return result.concat(left).concat(right);
};

let arr = [12, 35, 21, 1, 23, 3, 2, 4, 5, 13, 8, 9, 21];
console.log(mergeSort(arr));

File: 2.linked-list/single-linked-list/merge-sort/merge_sort_general.rb
def merge_sort(arr)
  if arr.length <= 1
    return arr
  end
  len = arr.length
  mid = ((0 + len - 1) / 2).round()
  left = arr[0..mid]
  right = arr[mid + 1..-1]
  return merge(merge_sort(left), merge_sort(right))
end

def merge(left, right)
  result = []
  while (left.length >= 1 and right.length >= 1)
    if (left[0] < right[0])
      result.push(left.shift)
    else
      result.push(right.shift)
    end
  end
  return result + left + right
end

arr = [1, 5, 6, 9, 4, 3, 11, 26, 85, 1, 0, 574, 799]
p merge_sort(arr)
